engine/lib/phx/src/engine/event_bus.rs

```diff
@@ -0,0 +1,318 @@
+        Some(self.cmp(other))
+    }
+}
+
+pub struct EventBus {
+    events: HashMap<String, Event>,
+    update_pass_map: HashMap<UpdatePass, BinaryHeap<MessageRequest>>,
+    next_tunnel_id: AtomicU32,
+    max_priority_locked: bool,
+}
+
+impl EventBus {
+    pub fn new() -> Self {
+        Self {
+            next_tunnel_id: AtomicU32::new(0),
+            update_pass_map: HashMap::new(),
+            events: HashMap::new(),
+            max_priority_locked: false,
+        }
+    }
+
+    pub fn lock_max_priority(&mut self) {
+        self.max_priority_locked = true;
+    }
+
+    pub fn dispatch(&mut self, update_pass: UpdatePass, lua: &mut Ref<Lua>) {
+        //* Can we handle the lua state differently? */
+        //* Maybe also with future potential multiple lua states in mind */
+        let globals = lua.globals();
+        let event_tunnels: Table = globals.get("EventTunnels").expect("Unknown table");
+
+        if let Some(message_heap) = self.update_pass_map.get_mut(&update_pass) {
+            let mut message_requests: Vec<_> = message_heap.drain().collect();
+
+            for message_request in message_requests.iter_mut() {
+                if let Some(event) = self.events.get(&message_request.name) {
+                    for subscriber in event.subscribers.iter() {
+                        let id = subscriber.tunnel_id;
+
+                        if !message_request.completed && !message_request.stay_alive {
+                            println!(
+                                "Dispatched event {} through tunnel with id {}",
+                                message_request.name, id
+                            )
+                        }
+
+                        // todo: only dispatch to subscriber with set entity_id if defined
+                        let tunnel_func: Function = event_tunnels
+                            .get(id)
+                            .expect(&format!("Unknown tunnel with id: {}", id));
+
+                        if let Err(e) = tunnel_func.call::<_, ()>(()) {
+                            trace!("{}", e);
+                        } else {
+                            message_request.completed = true; // Mark as completed
+                        }
+                    }
+                } else {
+                    panic!("Event not found: {}", message_request.name);
+                }
+            }
+
+            for message_request in message_requests {
+                if message_request.stay_alive || !message_request.completed {
+                    message_heap.push(message_request);
+                }
+            }
+        }
+    }
+
+    pub fn dispatch_all(&mut self, mut lua: Ref<Lua>) {
+        for update_pass in UpdatePass::iter() {
+            self.dispatch(update_pass, lua.borrow_mut());
+        }
+    }
+}
+
+#[luajit_ffi_gen::luajit_ffi]
+impl EventBus {
+    pub fn register(
+        &mut self,
+        event_name: String,
```

Comment:
ok.

Most event bus methods are called on startup or unloading so performance shouldn´t be an issue here. I don't want to loose the string event name here since it makes identification what everything does rather easy. One option could be to have a Lua enum here which holds the integer (but is it worth the interrupting the developer workflow when registering a new event? :D)

Reply:
If it's more convenient to use strings and doesn't hit performance then it should be ok.